crates/red_knot_python_semantic/src/types/property_tests.rs

```diff
@@ -123,14 +124,62 @@ impl Arbitrary for Ty {
     }
 
     fn shrink(&self) -> Box<dyn Iterator<Item = Self>> {
-        // This is incredibly naive. We can do much better here by
-        // trying various subsets of the elements in unions, tuples,
-        // and intersections. For now, we only try to shrink by
-        // reducing unions/tuples/intersections to a single element.
         match self.clone() {
-            Ty::Union(types) => Box::new(types.into_iter()),
-            Ty::Tuple(types) => Box::new(types.into_iter()),
-            Ty::Intersection { pos, neg } => Box::new(pos.into_iter().chain(neg)),
+            Ty::Union(types) => Box::new(types.shrink().filter_map(|elts| match elts.len() {
+                0 => None,
+                1 => Some(elts[0].clone()),
+                _ => Some(Ty::Union(elts)),
+            })),
+            Ty::Tuple(types) => Box::new(types.shrink().filter_map(|elts| match elts.len() {
+                0 => None,
+                1 => Some(elts[0].clone()),
+                _ => Some(Ty::Tuple(elts)),
+            })),
+            Ty::Intersection { pos, neg } => {
+                // Shrinking on intersections is not exhaustive!
+                //
+                // We try to shrink the positive side or the negative side,
+                // but we aren't shrinking both at the same time.
+                //
+                // This should remove positive or negative constraints but
+                // won't shrink (A & B & ~C & ~D) to (A & ~C) in one shrink
+                // iteration.
+                //
+                // Instead, it hopes that (A & B & ~C) or (A & ~C & ~D) fails
+                // so that shrinking can happen there.
+                let pos_orig = pos.clone();
+                let neg_orig = neg.clone();
+                Box::new(
+                    chain!(
+                        // we shrink negative constraints first, as
+                        // intersections with only negative constraints are
+                        // more confusing
+                        neg.shrink().map(move |shrunk_neg| Ty::Intersection {
+                            pos: pos_orig.clone(),
+                            neg: shrunk_neg,
+                        }),
+                        pos.shrink().map(move |shrunk_pos| Ty::Intersection {
+                            pos: shrunk_pos,
+                            neg: neg_orig.clone(),
+                        }),
+                    )
```

Comment:
Nit: Can you tell me more about what the benefits of the `chain` macro are over the `Iterator::chain` method?
```suggestion

                    // we shrink negative constraints first, as
                    // intersections with only negative constraints are
                    // more confusing
                    neg.shrink().map(move |shrunk_neg| Ty::Intersection {
                        pos: pos_orig.clone(),
                        neg: shrunk_neg,
                    }).chain(pos.shrink().map(move |shrunk_pos| Ty::Intersection {
                        pos: shrunk_pos,
                        neg: neg_orig.clone(),
                    })
```

I'd otherwise prefer the `std` chain method because it's what most people are familiar with

Reply:
it's easier to swap around elements of `chain!` ad-hoc, since they're on the same level of the AST.

 I also think aesthetically that it makes sense for the components of the chain to be at the same level, rather than the second one to be chained off of the first. But that's _purely_ an aesthetics thing.

I'll switch over to `a.chain(b)`.